Document the Dotty rewrite rules for migration (#20)
scalacenter/scala-3-migration-guide: docs/dotty-rewrites.md

# Dotty rewrite rules

## Usage

The Dotty compiler, named `dotc`, has been carefully designed to ease the migration from Scala 2.13 to Scala 3.0. It comes with a handful of utilities to encourage you to cross-compile your codebase:

### Migration mode

The `-source:3.0-migration` option makes the compiler forgiving on most of the dropped features, printing warnings in place of errors. Each warning is a strong indication that the the compiler is capable of safely rewriting the deprecated code into a cross-compiling code.

### Error explanations

In case the compiler does not handle some old construct, you can ask for explanation with the `-explain` option. This option is not limited to migration, it is, in general, a wonderful support to learn and code in Scala 3.

In combination with `-source:3.0-migration` the compiler indicates you how to fix the remaining errors by hand. Otherwise it won't be able to apply the rewrite rules.

### Automatic rewrites

> Once your code compiles with the `-source:3.0-migration`, all warnings can be resolved automatically by the addition of the `-rewrite` option. Beware that the compiler will modify your code! It is intended to be safe. However you may like to commit the initial state so that you can print the diff applied by the compiler and revert if necessary.

Author: @julienrf 

The answer of your first question is supposed to be implied in those lines. But maybe it should be underpinned...
On the other hand, the step by step tutorial will be a lot more explicit on this:
- first fix the errors
- then run the `-rewrite`

> Once your code compiles with the `-source:3.0-migration`, all warnings can be resolved automatically by the addition of the `-rewrite` option. Beware that the compiler will modify your code! It is intended to be safe. However you may like to commit the initial state so that you can print the diff applied by the compiler and revert if necessary.

Author: @julienrf 

The answer of your first question is supposed to be implied in those lines. But maybe it should be underpinned...
On the other hand, the step by step tutorial will be a lot more explicit on this:
- first fix the errors
- then run the `-rewrite`

Reviewer: I’m not sure it is covered: what I meant is what if my code contains type errors that the compiler doesn’t know how to fix?

Maybe the beginning should be rephrased like so? (IIUC)

```suggestion

In combination with `-source:3.0-migration` the compiler requires you to fix the remaining errors by hand, otherwise it won't be able to apply the rewrite rules.

### Automatic rewrites

Once your code compiles with the `-source:3.0-migration`, all warnings can be resolved automatically by the addition of the `-rewrite` option. Beware that the compiler will modify your code! It is intended to be safe. However you may like to commit the initial state so that you can print the diff applied by the compiler and revert if necessary.
```

# Dotty migration rewrite rules

## Usage

The Dotty compiler, named `dotc`, has been carefully designed to ease the migration from Scala 2.13 to Scala 3. It comes with a handfull of utilities to encourage you to cross-compile your codebase:

### Migration mode

> The `-source:3.0-migration` option makes the compiler forgiving on most of the dropped features, printing warnings in place of errors. Each warning is a strong indication that the the compiler is capable of safely rewritting the deprecated code into a cross-compiling code.

Reviewer 2: ```suggestion
The `-source:3.0-migration` option makes the compiler forgiving on most of the dropped features, printing warnings in place of errors. Each warning is a strong indication that the the compiler is capable of safely rewriting the deprecated code into a cross-compiling code.
```

### Error explanations

In case the compiler does not handle some old construct, you can ask for explanation with the `-explain` option. This option is not limited to migration, it is, in general, a wonderful support to learn and code in Scala 3.

In combination with `-source:3.0-migration` the compiler indicates you how to fix the remaining errors by hand. Otherwise it won't be able to apply the rewrite rules.

### Automatic rewrites

> Once your code compiles with the `-source:3.0-migration`, all warnings can be resolved automatically by the addition of the `-rewrite` option. Beware that the compiler will modify your code! It is intended to be safe. However you may like to commit the initial state so that you can print the diff applied by the compiler and revert if willing.

Reviewer 2: ```suggestion
Once your code compiles with the `-source:3.0-migration`, all warnings can be resolved automatically by the addition of the `-rewrite` option. Beware that the compiler will modify your code! It is intended to be safe. However you may like to commit the initial state so that you can print the diff applied by the compiler and revert if necessary.
```

## Documentation

### Rule 1 - Treat keyword as identifier

> Dotty introduces a number of keywords among which some are restricted meaning you cannot use them as identifiers anymore.

Reviewer 2: ```suggestion
Dotty introduces a number of keywords, among which some are restricted and some are "soft" keywords, meaning you cannot use them as identifiers anymore.
```

Not all the [Dotty keywords](https://dotty.epfl.ch/docs/internals/syntax.html#keywords) are restricted, the ones that are said to be _soft_ can still be used as identifiers, for instance the `open` keyword.

For the matter of migration from Scala 2 to Scala 3, the keywords at stake are:

> - `do`

Reviewer 2: `do` has always been a keyword.

Perhaps you were confused with the fact that the `do..while` loop control structure has been dropped?

Author: yes right!

- `enum`
- `export`
- `given`

> - `return`

Reviewer 2: `return` has always been a keyword.

Perhaps you mixed up with the fact that non-local returns are dropped?

Author: I wanted to double check and then I forgot so thank you for catching those!

- `then`
- `=>>`
- `?=>`

> Hopefully there is a straightforward solution which is to backquote the identifiers that cannot be used anymore. In most cases the rewrite rule can do it for you. Otherwise the compiler will point you toward the solution.

Reviewer 2: ```suggestion
There is a straightforward solution which is to backquote the identifiers that cannot be used anymore. In most cases the rewrite rule can do it for you. Otherwise the compiler will point you toward the solution.
```

For instance the following Scala 2 code

```scala
object given {
  val enum = ???

  def foo = println(enum)
}
```

Compiled with `dotc -source:3.0-migration -rewrite` is rewritten into:

```scala
object `given` {
  val `enum` = ???

  def foo = println(`enum`)
}
```

In some cases the compiler is not able to apply the solution by itself but then you can add the `-explain` option and it will point you towards the solution.

With the following input source:

```scala
object Foo {
  def do: Unit = ()

> }

Reviewer 2: This is not valid Scala 2 code:
```
>scala
Welcome to Scala 2.13.2 (OpenJDK 64-Bit Server VM, Java 1.8.0_222).
Type in expressions for evaluation. Or try :help.

scala> object Foo { def do: Unit = () }
                        ^
       error: identifier expected but 'do' found.
                                     ^
       warning: procedure syntax is deprecated: instead, add `: Unit` to explicitly declare `<error>`'s return type
```

```

It prints this explanation:

```shell
-- [E040] Syntax Error: tests/rewrites-doc/treatAsIdent.scala:2:6 --------------
2 |  def do: Unit = ()
  |      ^^
  |      an identifier expected, but 'do' found

Explanation
===========

If you want to use 'do' as identifier, you may put it in backticks: `'do'`.
```

### Rule 2 - Add Parentheses around the parameter of a lambda

> A lambda parameter with type ascripion is now required to be enclosed in parentheses, making the following Scala 2 code illegal:

Reviewer 2: ```suggestion
A lambda parameter with an explicit type is now required to be enclosed in parentheses, making the following Scala 2 code illegal:
```
A type ascription is a term of the form `expr: T`. For formal parameters we talk about whether they have an explicit type or not.

Author: I dont get it.

What is a formal parameter? How is it different to a term of the form `expr: T`?

```scala
object Foo {
  val f = { x: Int => x * x }
}
```

Compiling with `dotc -source:3.0-migration -rewrite` rewrites it into:

```scala
object Foo {
  val f = { (x: Int) => x * x }
}
```

### Rule 3 - Replace symbol literals by `Symbol` application

> Symbol literals are no longer supported, but the `scala.Symbol` class still exists so that each string litteral can be safely replaced by an application of `Symbol`.

Reviewer 2: It's perhaps worth pointing out that symbol literals are deprecated in Scala 2.13.3+.

> Although `scala.Symbol` is usefull for migration, beware that it will be soon deprecated and removed from `scala-library`. You are recommended, as a second step, to replace them with plain string literals `"xwy"` or quoted expressions ``'{xyz}``.

Reviewer 2: ```suggestion
Although `scala.Symbol` is useful for migration, beware that it will be soon deprecated and removed from `scala-library`. You are recommended, as a second step, to replace them with plain string literals `"xwy"` or a dedicated class.
```
Don't mention quoted expressions here. They are used for metaprogramming. It's an entirely different world than symbol literals.

Author: I was confused because of the compiler message but now I get it :)
```
  |       symbol literal 'abc is no longer supported,
  |       use a string literal "abc" or an application Symbol("abc") instead,
  |       or enclose in braces '{abc} if you want a quoted expression.
```

In the following example:

```scala
object Foo {
  val values: Map[Symbol, Int] = Map('abc -> 1)

  val abc = values('abc)
}
```

Compiling with `dotc -source:3.0-migration -rewrite` rewrites the code into:

```scala
object Foo {
  val values: Map[Symbol, Int] = Map(Symbol("abc") -> 1)

  val abc = values(Symbol("abc"))
}
```

### Rule 4 - Indent open brace when passing an argument

In Scala 2 it is possible to pass an argument after a new line by enclosing it into braces. Although valid, this style of coding is not encouraged by the [Scala style guide](https://docs.scala-lang.org/style) and it is no longer supported in Scala 3.

Example:
```scala
object Foo {
  def test(title: String)(body: => Unit): Unit = {
    println(title)
    body
  }

  test("my test")
  {
    assert(1 == 1)
  }
}
```

The solution is to indent the block, or at least the first line. Compiling with `dotc -source:3.0-migration -rewrite` rewrites the code into:
```scala
object Foo {
  def test(title: String)(body: => Unit): Unit = {
    println(title)
    body
  }

  test("my test")

>     {

Reviewer 3: are these misaligned braces inserted by dotty?

Reviewer 2: Yes, unfortunately.

IMO Dotty should move the brace at the end of the previous line instead:
```scala
  test("my test") {
    assert(1 == 1)
  }
```

Author: Yes it is!

Dotty indent the first brace only to make the code compile.

It would not be hard to change this rule to write that instead: 
```scala
test("my test") {
 <body>
}
```

    assert(1 == 1)
  }
}
```

> Even though the style of the resulting code is not elegant it compiles, helping you to make the transition to a different style.

Reviewer 2: ```suggestion
Even though the style of the resulting code is not elegant, it compiles, helping you to make the transition to a different style.
```

This rule applies to other patterns as well, such as refining a type after a new line. The resulting code looks like:

```scala
trait Foo

object Foo {
  type Bar = Foo

>    {

Reviewer 2: ```suggestion
    {
```
?

    def bar(): Int
  }
}
```

### Rule 5 - Replace `do`-`while` construct with `while` equivalent

The `do` keyword has acquired a different meaning in the [new syntax rules](https://dotty.epfl.ch/docs/reference/other-new-features/control-syntax).

> To avoid confusion the traditional `do <body> while (<cond>)` construct is dropped. It is recommended to use the equivalent `while ({ <body>; <cond> }) ()` that cross-compiles or the new Scala 3 syntax `while { <body>; <cond> } do ()`.

Reviewer 2: ```suggestion
To avoid confusion, the traditional `do <body> while (<cond>)` construct is dropped. It is recommended to use the equivalent `while ({ <body>; <cond> }) ()` that cross-compiles or the new Scala 3 syntax `while { <body>; <cond> } do ()`.
```

For instance, compiling the following code with `dotc -source:3.0-migration -rewrite` 

```scala

> do
>   i += 1
> while (f(i) == 0)

Reviewer 2: The standard way of writing this in Scala 2 would involve braces:
```suggestion
do {
  i += 1
} while (f(i) == 0)
```

Author: Agreed!

After trying I get this result:
```scala
while ({ {
      i += 1
    } ; f(i) == 0}) ()
```
It looks ugly to me but maybe it is hard to improve or it is not worth it...

```

rewrites it into

```scala
while ({
  i += 1
  f(i) == 0
}) ()
```

For more information you may refer to [this page](https://dotty.epfl.ch/docs/reference/dropped-features/do-while.html).

### Rule 6 - Drop procedure syntax

> Procedure syntax has been dropped, making the following Scala 2 code illegal:

Reviewer 2: It's perhaps worth pointing out that it's been deprecated in Scala 2 for a while already.

```scala
trait Foo {
  def print()
}

object Bar {
  def print() {
    println("bar")
  }
}
```

Compiling with `dotc -source:3.0-migration -rewrite` rewrites it into:

```scala
trait Foo {
  def print(): Unit 
}

object Bar {
  def print(): Unit = {
    println("bar")
  }
}
```

## Looking ahead to Scala 3.1

> Some deprecations have been postpone to 3.1 to facilitate the migration from 2.13 to 3.0 and then from 3.0 to 3.1. However some of the migration rules are already available and you are likely to being able to apply them in your codebase. In this way you get accustom to the new syntax and prepared for 3.1.

Reviewer 2: ```suggestion
Some deprecations have been postponed to 3.1 to facilitate the migration from 2.13 to 3.0 and then from 3.0 to 3.1. However some of the migration rules are already available and you are likely to be able to apply them in your codebase. In this way you get accustomed to the new syntax and prepared for 3.1.
```

### Rule 7 - Replace wildcard type argument '_' with '?'

The deprecation of the `_` syntax for wildcard type arguments has be postoned to Scala 3.1 and it will be replaced by the `?` syntax. You can find the motivation for this in the [Dotty reference](https://dotty.epfl.ch/docs/reference/changed-features/wildcards.html).

> However there already is a migration rule that you can apply using some slightly different options than before: `dotc -source:3.1-migration -rewrite`. Note the `3.1` version mentionned here.

Reviewer 2: ```suggestion
However there already is a migration rule that you can apply using some slightly different options than before: `dotc -source:3.1-migration -rewrite`. Note the `3.1` version mentioned here.
```

# Dotty rewrite rules

## Usage

The Dotty compiler, named `dotc`, has been carefully designed to ease the migration from Scala 2.13 to Scala 3.0. It comes with a handful of utilities to encourage you to cross-compile your codebase:

### Migration mode

The `-source:3.0-migration` option makes the compiler forgiving on most of the dropped features, printing warnings in place of errors. Each warning is a strong indication that the the compiler is capable of safely rewriting the deprecated code into a cross-compiling code.

### Error explanations

In case the compiler does not handle some old construct, you can ask for explanation with the `-explain` option. This option is not limited to migration, it is, in general, a wonderful support to learn and code in Scala 3.

In combination with `-source:3.0-migration` the compiler indicates you how to fix the remaining errors by hand. Otherwise it won't be able to apply the rewrite rules.

### Automatic rewrites

> Once your code compiles with the `-source:3.0-migration`, all warnings can be resolved automatically by the addition of the `-rewrite` option. Beware that the compiler will modify your code! It is intended to be safe. However you may like to commit the initial state so that you can print the diff applied by the compiler and revert if necessary.

Author: @julienrf 

The answer of your first question is supposed to be implied in those lines. But maybe it should be underpinned...
On the other hand, the step by step tutorial will be a lot more explicit on this:
- first fix the errors
- then run the `-rewrite`

## Documentation

> ### Rule 1 - Treat keyword as identifier

Reviewer: ```suggestion
### Rule 1 - Treat keywords used as identifiers
```

Or maybe?

```suggestion
### Rule 1 - Escape keywords used as identifiers
```

- `enum`
- `export`
- `given`
- `then`
- `=>>`
- `?=>`

There is a straightforward solution which is to backquote them. In most cases the rewrite rule can do it for you. Otherwise the compiler will point you toward the solution.

For instance the following Scala 2 code:

```scala
object given {
  val enum = ???

  def foo = println(enum)
}
```

Compiled with `dotc -source:3.0-migration -rewrite` is rewriten into:

```scala
object `given` {
  val `enum` = ???

  def foo = println(`enum`)
}
```

In some cases the compiler is not able to apply the solution by itself but then you can add the `-explain` option and it will point you towards the solution.

With the following input source:

```scala
object Foo {
  def then(): Unit = ()
}
```

It prints this explanation:

```shell
-- [E040] Syntax Error: tests/rewrites-doc/1-treatAsIdent.scala:2:6 ------------
2 |  def then(): Unit = ()
  |      ^^^^
  |      an identifier expected, but 'then' found

Explanation
===========

If you want to use 'then' as identifier, you may put it in backticks: `then`.
```

### Rule 2 - Add Parentheses around the parameter of a lambda

> A lambda parameter with type ascription is now required to be enclosed in parentheses, making the following Scala 2 code illegal:

Reviewer 2: Apparently my earlier comment on this was lost:
```suggestion
A lambda parameter with an explicit type is now required to be enclosed in parentheses, making the following Scala 2 code illegal:
```

```scala

> object Foo {
>   val f = { x: Int => x * x }
> }

Reviewer: I would simplify to:

```suggestion
val f = { x: Int => x * x }
```

So that we focus only on the relevant parts (although this piece of code is not a valid Scala 2 file)

Reviewer: I would simplify the other code examples in the same way.

```scala
trait Foo {
  def print(): Unit 
}

object Bar {
  def print(): Unit = {
    println("bar")
  }
}
```

## Looking ahead to Scala 3.1

Some deprecations have been postponed to 3.1 to facilitate the migration from 2.13 to 3.0 and then from 3.0 to 3.1. However some of the migration rules are already available and you are likely to be able to apply them in your codebase. In this way you get accustomed to the new syntax and prepared for 3.1.

### Rule 7 - Replace wildcard type argument '_' with '?'

The deprecation of the `_` syntax for wildcard type arguments has be postponed to Scala 3.1 and it will be replaced by the `?` syntax. You can find the motivation for this in the [Dotty reference](https://dotty.epfl.ch/docs/reference/changed-features/wildcards.html).

However there already is a migration rule that you can apply using some slightly different options than before: `dotc -source:3.1-migration -rewrite`. Note the `3.1` version mentioned here.

For example:

```scala
trait Container[A] {
  def weight: Int
}

object Container {
  def compare(x: Container[_], y: Container[_]): Int = {
    x.weight - y.weight
  }
}
```

> Is rewriten into

Reviewer: ```suggestion
Is rewritten into
```
